### DISCUSSION TITLE
Sync_items scalability question

### DISCUSSION THREAD

**Discussion by mautomic** [ecederstrand/exchangelib#922] [Q&A]
Hi all, 

I recently started utilizing the synchronization functionality that exchangelib has to offer, and it's working out quite well. However my general concern is if it's scalable, and curious on how others are handling using it with very large inboxes. 

My application only needs to listen for new emails or updates on emails from the current point in time onwards, so to run a 'baseline sync' when the application launches, I started off with something like this (very primitive):

```python3
for _, item in folder.sync_items():
    pass
return folder.item_sync_state
```

This sets the initial sync state which I save, and then at a periodic interval I make `sync_items` calls for new events.

This did the job until I ran into an inbox that contained over 40k emails, and it took ~7 minutes to complete. So I looked into the functionality of `sync_items` and made a couple optimizations to `max_changes_returned` (default chunking is only 100) and `only_fields` (getting all fields by default).

```python3
for _, item in folder.sync_items(max_changes_returned=10000, only_fields=[]):
    pass
return folder.item_sync_state
```

Now this helped reduce the time to sync a large inbox from ~7 minutes to about ~20-30 seconds. 

I was curious if there are other ways to optimize with this functionality, or if there are any other suggestions on being more scalable? I can move forward with 30 second syncing, but would prefer to get it down to under 10 seconds. 

**One feature I was thinking of (and not sure if possible), is if we can artificially acquire/create a sync state without actually making the API calls to EWS and not yielding items? This could be a flag that you'd pass in that would indicate starting from the current point in time, and it would immediately return a sync state.**

I haven't done much digging yet, but if this sounds feasible, I would love to help contribute to this awesome library! If something like this already exists or can already be achieved, and I've missed it, apologies. 

Cheers

**Comment by ecederstrand** [comment]
If you're just interested in changes from the current point in time and don't care about historical changes, couldn't you create a streaming subscription and use `get_streaming_events()` to fetch the changes?

```python
with a.inbox.streaming_subscription() as subscription_id:
    for notification in a.inbox.get_streaming_events(subscription_id):
        for event in notification.events:
            # Do something
```

**Reply by mautomic** [reply]
This could be worth looking at for us - but does the subscription have to be long running/maintained? 

Our application utilizes distributed task queues, so we launch a periodical ephemeral task that will "ingest emails", which can just lookup the previous sync_state that we save in our database from the last run, and use that to check for events. Hence we believed that synchronization was the best fit for our architecture. 

Technically, I guess I could launch a single task that runs as long as the application is running, who's sole purpose is for streaming, but that would require a discussion between me and my team :)

**Reply by mautomic** [reply]
@ecederstrand, does my idea outlined above in the initial post sound possible? - or does EWS limit the flexibility on this, regarding how we get the sync_state back.

**Reply by ecederstrand** [reply]
In that case no, streaming events is not optimal. This synchronization type is intended for long-running processes (minutes, not seconds). For cron-like jobs, `sync_items()` is better. If you don't want to start from the beginning every time, you can save the sync state and use that as input for the next `sync_items()` call:

```python
# Initial
for change_type, item in a.inbox.sync_items():
    # Do something

# Store this somewhere
state = item.item_sync_state

# Next time you want to sync
for change_type, item in a.inbox.sync_items(sync_state=state):
    # Do something

```

**Reply by mautomic** [reply]
Ah yes, this is what we're doing - I'm curious on optimization on that very first sync.

**Reply by ecederstrand** [reply]
Yeah, sorry, you already mentioned `item_sync_state` in your initial post. I don't think you can have it both ways - you have to replay all events at least once with `sync_items()` to get the sync state, or use the streaming option.

**Reply by mautomic** [reply]
Got it, makes sense. I think setting those two params mostly did the trick in improving performance. Any other ideas? It's possible we're reaching the upper bound now and I'm just nitpicking :)

Anyways, thanks for taking the time to answer! Truly appreciate it.

**Reply by ecederstrand** [reply]
If you haven't seen it already, https://ecederstrand.github.io/exchangelib/#synchronization-subscriptions-and-notifications has links to MSDN articles discussing the different sync strategies in EWS.
